### PR TITLE
Return to ground when we flush the last char

### LINKED ISSUE #2746: Alt-Backspace followed by Enter does not behave properly
Pressing Alt+Backspace, then some other characters, followed by Enter will not trigger the right inputs for a shell through WSL. This issue is pretty subtle and it can be tricky to observe in my experience, so I have 2 test cases below to demonstrate this issue.

# Environment

```none
Windows build number: 10.0.18362.0
Windows Terminal version (if applicable): 0.4.2382.0 (latest available on the Windows Store)

Any other software? Ubuntu 18.04.3 LTS in WSL 1; Linux Mint with GNOME Terminal 3.28.1
```

# Test case 1

## Steps to reproduce

1. Open a Bash terminal
2. Type `echo foo` (don't press Enter)
3. Press Alt+Backspace to delete the word (the command should now read `echo`)
4. Type `bar` (the command should now read `echo bar`)
5. Press Enter

## Expected behavior

`bar` should be printed. This is the behavior when using WSL via vanilla Command Prompt, or using GNOME Terminal on a standalone Linux distro.

## Actual behavior

Nothing happens. Pressing Alt+Backspace then pressing Enter inputs an escape sequence that would be expected from Alt+Enter, which Bash silently ignores by default.

To see this more visibly, run `bind -x '"\e\r":"echo whoops"'` then repeat the above steps. Now, `whoops` will be printed, since the above steps will input the escape sequence for Alt+Enter (`\e\r`).

# Test case 2

## Steps to reproduce

1. Open a Bash terminal
2. Run `cat > sample.txt`
3. Type `foo` followed by Alt+Backspace followed by `bar` followed by Enter
4. Press Ctrl+D to return to the prompt
5. Run `hexdump -C sample.txt`

## Expected behavior

`hexdump` should print the following output:

```
00000000  66 6f 6f 62 61 72 0a                              |foobar.|
00000007
```

This is the behavior when using WSL via vanilla Command Prompt, or using GNOME Terminal on a standalone Linux distro.

## Actual behavior

`hexdump` prints the following output:

```
00000000  66 6f 6f 62 61 72 1b 0a                           |foobar..|
00000008
```

Just like before, we can see the sequence for Alt+Enter (`\e\r`, or `1b 0a` in hex) is actually registered where we expected to see just `\r` (`0a` in hex).

### PR DESCRIPTION
**Pull request by zadjii-msft**
## Summary of the Pull Request
The InputStateMachineEngine was incorrectly not returning to the ground state after flushing the last sequence. That means that something like alt+backspace would leave us in the Escape state, not the ground state. This makes sure we return to ground.

Additionally removes the "Parser.UnitTests-common.vcxproj" file, which was originally used for a theoretical time when we only open-sourced the parser. It's unnecessary now, and we can get rid of it.

Also includes a small patch to bcz.cmd, to make sure bx works with projects with a space in their name.

## PR Checklist
* [x] Closes #2746
* [x] I work here
* [x] Tests added/passed
* [n/a] Requires documentation to be updated

### CODE REVIEW

**Comment by zadjii-msft** [comment]
@DHowett-MSFT None of the `Action*` methods modify the state, they only ever call the appropriate methods on the engine with the appropriate args. Typically, the `Event*` methods are the ones that actually modify the state. This block is basically just the collected `Dispatch` parts of each of those states, and it should definitely have this. I can't think of a case where when we flush like this, we'd want to leave it in the not-ground state. The whole point of the flushing thing is to be able to handle weird partial state VT input sequences, like `\x1b\x7f` or `\x1b\x1b`, that would normally not actually get dispatched on their own and would pollute the parser state like this.

File: src/terminal/parser/stateMachine.cpp [hunk 1]
```diff
@@ -1413,25 +1413,31 @@ void StateMachine::ProcessString(const wchar_t* const rgwch, const size_t cch)
...
             case VTStates::Ss3Entry:
             case VTStates::Ss3Param:
-                return _ActionSs3Dispatch(*pwch);
-            default:
-                return;
+                _ActionSs3Dispatch(*pwch);
+                break;
             }
+            // microsoft/terminal#2746: Make sure to return to the ground state
+            // after dispating the characters
```
Comment (DHowett-MSFT) [inline src/terminal/parser/stateMachine.cpp]:
```suggestion
            // after dispatching the characters
```

**Comment by DHowett-MSFT** [comment]
How will this impact partial writing of VT sequences (`\x1b[1` ... `;1H`)? Was that something we supported before?

**Comment by zadjii-msft** [comment]
@DHowett-MSFT Fortunately, that still works for the OutputStateMachineEngine. Each engine implements `FlushAtEndOfString` differently, so the OutputEngine lets the parser stay buffered, while this change _only_ applies to the InputEngine

**Comment by DHowett-MSFT** [comment]
Hmm. _hmm_. I was thinking that this might be fine since it's on Input (my bad: i forgot this wasn't output), but consider the case where SSH is buffering and the user holds down <kbd>right arrow</kbd> a whole bunch. Maybe it gets a sequence split across a packet?

File: tools/bcz.cmd [hunk 2]
```diff
@@ -68,7 +68,7 @@ if "%_EXCLUSIVE%" == "1" (
 echo Performing nuget restore...
 nuget.exe restore %OPENCON%\OpenConsole.sln
 
-set _BUILD_CMDLINE="%MSBUILD%" %OPENCON%\OpenConsole.sln /t:%_MSBUILD_TARGET% /m /p:Configuration=%_LAST_BUILD_CONF% /p:Platform=%ARCH% %_APPX_ARGS%
+set _BUILD_CMDLINE="%MSBUILD%" %OPENCON%\OpenConsole.sln /t:"%_MSBUILD_TARGET%" /m /p:Configuration=%_LAST_BUILD_CONF% /p:Platform=%ARCH% %_APPX_ARGS%
```
Comment (carlos-zamora) [inline tools/bcz.cmd]: wut? why?

Comment (zadjii-msft) [reply tools/bcz.cmd]: > The _MSBUILD_TARGET can have spaces in it, and if it does, this won't work.Case in point: any project in the "Virtual Terminal" solution directory. Wrapping it with quotes fixes that.

Comment (zadjii-msft) [reply tools/bcz.cmd]: I _knew_ I had this in another PR somewhere, but it _shouldn't_ conflict...

**Comment by miniksa** [comment]
> > Can you clarify why manually executing these things for the input machine is required? I traced around to your comments on InputStateMachineEngine::FlushAtEndOfString() and I'm still unclear why this is a special custom case here.
> 
> Sure thing.
> 
> One of the "weird things" in VT input is the case of something like alt+[. In VT, that's encoded as `\x1b[`. However, that's also the start of a CSI, and could be the start of a longer sequence, there's no way to know for sure. For an alt+[ keypress, the parser originally would just sit in the `CsiEntry` state after processing it, which would pollute the following keypress (e.g. alt+[, A would be processed like `\x1b[A`, which is _wrong_).
> 
> Fortunately, for VT input, each keystroke comes in as an individual write operation. So, if at the end of processing a string for the InputEngine, we find that we're not in the Ground state, that implies that we've processed some input, but not dispatched it yet. This block at the end of `ProcessString` will then re-process the undispatched string, but it will ensure that it dispatches on the last character of the string. For our previous `\x1b[` scenario, that means we'll make sure to call `_ActionEscDispatch('[')`., which will properly decode the string as alt+[.

Can you leave some portion of this clarification in the code right there as a comment?

It's way clearer when you explain the Alt+[ causing it to sit in the wrong state than it is by simple inspection.

**Comment by miniksa** [comment]
> add the comment @miniksa wanted

IT IS BEAUTIFUL. Thanks.

**Pull request merged**